### PR TITLE
Allow skipping forks and mirrors from being indexed

### PR DESCRIPTION
**Pull request by techknowlogick**
This PR adds two new options to disable repo/code search indexing of both forks and mirrors.

Related: #22842

### CODE REVIEW

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -1384,6 +1384,12 @@ ROUTER = console
 ;; repo indexer by default disabled, since it uses a lot of disk space
 ;REPO_INDEXER_ENABLED = false
 ;;
+;; repo indexer skip forks from being indexed if setting is enabled
+;REPO_INDEXER_SKIP_FORKS = false
+;;
+;; repo indexer skip mirrors from being indexed if setting is enabled
+;REPO_INDEXER_SKIP_MIRRORS = false
```
Comment (delvh) [inline custom/conf/app.example.ini]: Hmm…
I don't think this is the best approach.
What I imagine to be much more future proof would be something like
```ini
;; Define repo types that won't be indexed by the repo indexer
;; Possible values are fork, mirror, [in the future] username-matches:$REGEX, reponame-matches:$REGEX
;REPO_INDEXER_SKIPPED_REPOS = fork, mirror, [in the future] username-matches:.*test-user, reponame-matches:.*repo-?\d*
```
so, for the current state:
```suggestion
;; Define repo types that won't be indexed by the repo indexer
;; Possible values are fork and mirror
;REPO_INDEXER_SKIPPED_REPOS =
```

Comment (lunny) [reply custom/conf/app.example.ini]: Mixing types and name matching looks like not a good idea. Maybe one option for types, one for name matching.

**Comment by techknowlogick** [comment]
@delvh / @lunny I've updated this PR, I'm also hesitant about mixing types and name matching. I am also hesitant about regex, although that could easily be solved with globbing.

This at least reduces multiple options that I had originally introduced as a part of this PR

**Comment by lunny** [comment]
> @delvh / @lunny I've updated this PR, I'm also hesitant about mixing types and name matching. I am also hesitant about regex, although that could easily be solved with globbing.
> 
> This at least reduces multiple options that I had originally introduced as a part of this PR

How about just putting types, now we have only `original`, `forks`, `mirrors`, `templates`?

File: custom/conf/app.example.ini [hunk 2]
```diff
@@ -1380,6 +1380,9 @@ ROUTER = console
 ;; repo indexer by default disabled, since it uses a lot of disk space
 ;REPO_INDEXER_ENABLED = false
 ;;
+;; repo indexer units, the items to index, could be `repos`, `forks`, `mirrors` or any combination of them separated by a comma
+;REPO_INDEXER_UNITS = repos,forks,mirrors
```
Comment (lunny) [inline custom/conf/app.example.ini]: How about name it `REPO_INDEXER_TYPES` or `REPO_INDEXER_REPO_TYPES`?

**Comment by techknowlogick** [comment]
@lunny ok. I've updated with your feedback.

File: docs/content/doc/administration/config-cheat-sheet.en-us.md [hunk 3]
```diff
@@ -465,6 +465,7 @@ relation to port exhaustion.
 - `ISSUE_INDEXER_PATH`: **indexers/issues.bleve**: Index file used for issue search; available when ISSUE_INDEXER_TYPE is bleve and elasticsearch. Relative paths will be made absolute against _`AppWorkPath`_.
 
 - `REPO_INDEXER_ENABLED`: **false**: Enables code search (uses a lot of disk space, about 6 times more than the repository size).
+- `REPO_INDEXER_UNITS`: **sources,forks,mirrors,templates**: Repo indexer units. The items to index could be `repos`, `forks`, `mirrors`, `templates` or any combination of them separated by a comma. If empty then it defaults to `sources` only, as if you'd like to disable fully please see `REPO_INDEXER_ENABLED`.
```
Comment (wxiaoguang) [inline docs/content/doc/administration/config-cheat-sheet.en-us.md]: The ini says: `;REPO_INDEXER_REPO_TYPES = sources,forks,mirrors,templates`

Here says: `repos`, `forks`, `mirrors`, `templates`

File: modules/setting/indexer.go [hunk 4]
```diff
@@ -40,6 +41,7 @@ var Indexer = struct {
 	IssueIndexerName: "gitea_issues",
 
 	RepoIndexerEnabled: false,
+	RepoIndexerUnits:   []string{"repo", "fork", "mirror"},
```
Comment (wxiaoguang) [inline modules/setting/indexer.go]: And here is `	RepoIndexerUnits:   []string{"repo", "fork", "mirror"},`

Three kinds of concepts?

Comment (wxiaoguang) [reply modules/setting/indexer.go]: I guess it needs some constants ......

File: modules/setting/indexer.go [hunk 5]
```diff
@@ -40,6 +41,7 @@ var Indexer = struct {
 	IssueIndexerName: "gitea_issues",
 
 	RepoIndexerEnabled: false,
+	RepoIndexerUnits:   []string{"sources", "fork", "mirror", "templates"},
```
Comment (wxiaoguang) [inline modules/setting/indexer.go]: Hmm ... it still doesn't match.  `fork` vs `forks`, `mirror` vs `mirrors`.

File: modules/setting/indexer.go [hunk 6]
```diff
@@ -71,6 +73,7 @@ func loadIndexerFrom(rootCfg ConfigProvider) {
 	Indexer.IssueIndexerName = sec.Key("ISSUE_INDEXER_NAME").MustString(Indexer.IssueIndexerName)
 
 	Indexer.RepoIndexerEnabled = sec.Key("REPO_INDEXER_ENABLED").MustBool(false)
+	Indexer.RepoIndexerUnits = strings.Split(sec.Key("REPO_INDEXER_REPO_TYPES").MustString("sources,forks,mirrors,templates"), ",")
```
Comment (wxiaoguang) [inline modules/setting/indexer.go]: If I remember correctly, there is a function called `Strings` or something, which can trim spaces for every field?

Comment (techknowlogick) [reply modules/setting/indexer.go]: Yes, there is, but it doesn't work with "MustString" so you can't set a default with it.

Comment (wxiaoguang) [reply modules/setting/indexer.go]: The ini package is really problematic .... 😭

File: modules/setting/indexer.go [hunk 7]
```diff
@@ -40,6 +41,7 @@ var Indexer = struct {
 	IssueIndexerName: "gitea_issues",
 
 	RepoIndexerEnabled: false,
+	RepoIndexerUnits:   []string{"sources", "forks", "mirrors", "templates"},
```
Comment (lunny) [inline modules/setting/indexer.go]: The default value is only `sources` according `app.example.ini`

File: docs/content/doc/administration/config-cheat-sheet.en-us.md [hunk 8]
```diff
@@ -465,6 +465,7 @@ relation to port exhaustion.
 - `ISSUE_INDEXER_PATH`: **indexers/issues.bleve**: Index file used for issue search; available when ISSUE_INDEXER_TYPE is bleve and elasticsearch. Relative paths will be made absolute against _`AppWorkPath`_.
 
 - `REPO_INDEXER_ENABLED`: **false**: Enables code search (uses a lot of disk space, about 6 times more than the repository size).
+- `REPO_INDEXER_UNITS`: **sources,forks,mirrors,templates**: Repo indexer units. The items to index could be `sources`, `forks`, `mirrors`, `templates` or any combination of them separated by a comma. If empty then it defaults to `sources` only, as if you'd like to disable fully please see `REPO_INDEXER_ENABLED`.
```
Comment (lunny) [inline docs/content/doc/administration/config-cheat-sheet.en-us.md]: The default value is `sources,forks,mirrors,templates`?

Comment (techknowlogick) [reply docs/content/doc/administration/config-cheat-sheet.en-us.md]: yes

File: modules/setting/indexer.go (hunk 7 above)
Comment (techknowlogick) [reply modules/setting/indexer.go]: `app.example.ini` has all of them https://github.com/go-gitea/gitea/pull/23187/files#diff-eda333ab2034231553808ef0f99e28cd1f209c7f583fd3d0aa0fb1b77e573cd4R1386

**Review by lunny — APPROVED** [review]

File: modules/setting/indexer.go (hunk 6 above)
Comment (lunny) [inline modules/setting/indexer.go]:
```suggestion
	Indexer.RepoIndexerRepoTypes = strings.Split(sec.Key("REPO_INDEXER_REPO_TYPES").MustString("sources,forks,mirrors,templates"), ",")
```

Comment (techknowlogick) [reply modules/setting/indexer.go]: @lunny updated, but GitHub isn't showing the new commit yet.

**Review by wxiaoguang — APPROVED** [review]
IMO some nits :

* Maybe the `REPO_INDEXER_ENABLED` could be reused for `true = sources,forks,mirrors,templates`, then no need to introduce more and more new options
* The values like `sources`, `forks` are used in many places, to improve robustness, some constants could help.

Nits don't block, so vote my approval.

**Pull request merged**